File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/src/vector3D.cpp
/*
** EPITECH PROJECT, 2026
** snd_year_projects
** File description:
** vector3D
*/

#include "../include/vector3D.hpp"
using namespace Math;

Vector3D::Vector3D(double valX, double valY, double valZ)
    : X(valX), Y(valY), Z(valZ)
{}

double Vector3D::length()
{
    return sqrt(pow(X, 2) + pow(Y, 2) + pow(Z, 2));
}

double Vector3D::dot(const Vector3D& other)
{
    return (this->X * other.X) + (this->Y * other.Y) + (this->Z * other.Z);
}

double Vector3D::dot(const Vector3D& other) const
{
    return (this->X * other.X) + (this->Y * other.Y) + (this->Z * other.Z);
}

Vector3D Vector3D::operator+(const Vector3D& other)
{
    return Vector3D(this->X + other.X, this->Y + other.Y, this->Z + other.Z);
}

Vector3D Vector3D::operator+=(const Vector3D& other)
{
    Vector3D _new = *this + other;
    *this = _new;
    return *this;
}

Vector3D Vector3D::operator-=(const Vector3D& other)
{
    Vector3D _new = *this - other;
    *this = _new;
    return *this;
}

Vector3D Vector3D::operator-(const Vector3D& other)
{
    return Vector3D(this->X - other.X, this->Y - other.Y, this->Z - other.Z);
}

Vector3D Vector3D::operator*(double value)
{
    return Vector3D(this->X * value, this->Y * value, this->Z * value);
}

Vector3D Vector3D::operator*=(double value)
{
    Vector3D _new = *this * value;
    *this = _new;
    return *this;
}

Vector3D Vector3D::operator/(double value)
{
    return *this * (1 / value);
}

Vector3D Vector3D::operator/=(double value)
{
    Vector3D _new = *this / value;
    *this = _new;
    return *this;
}

Vector3D Vector3D::operator*(const Vector3D& other)
{
    return Vector3D(((Y * other.Z) - (Z * other.Y)), ((Z * other.X) - (X * other.Z)), ((X * other.Y) - (Y * other.X)));
}

Vector3D Vector3D::operator*=(const Vector3D& other)
{
    Vector3D _new = *this * other;
    *this = _new;
    return *this;
}

Vector3D Vector3D::operator/(const Vector3D& other)
{
    double x = 0.0;
    double y = 0.0;
    double z = 0.0;

    other.X == 0.0 ? x = 0.0 : x = this->X / other.X;
    other.Y == 0.0 ? y = 0.0 : y = this->Y / other.Y;
    other.Z == 0.0 ? z = 0.0 : z = this->Z / other.Z;

    return Vector3D(x, y, z);
}

Vector3D Vector3D::operator/=(const Vector3D& other)
{
    Vector3D _new = *this / other;
    *this = _new;
    return *this;
}

File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/rust/src/vector3d.rs
//
// EPITECH PROJECT, 2026
// snd_year_projects
// File description:
// vector3D
//

use std::ops;

#[derive(Debug, Clone, Copy)]
pub struct Vector3D {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl Vector3D {
    pub fn length(&self) -> f32 {
        return (self.x.powf(2.0) + self.y.powf(2.0) + self.z.powf(2.0)).sqrt();
    }

    pub fn dot(&self, other: Self) -> f32 {
        return (self.x * other.x) + (self.y * other.y) + (self.z * other.z);
    }
}

impl ops::Add<Vector3D> for Vector3D {
    type Output = Vector3D;

    fn add(self, other: Vector3D) -> Vector3D {
        return Vector3D { x: self.x + other.x, y: self.y + other.y, z: self.z + other.z };
    }
}

impl ops::Sub<Vector3D> for Vector3D {
   type Output = Vector3D;

    fn sub(self, other: Vector3D) -> Vector3D {
        return Vector3D { x: self.x - other.x, y: self.y - other.y, z: self.z - other.z };
    }
}

impl ops::Div<Vector3D> for Vector3D {
    type Output = Vector3D;

    fn div(self, other: Vector3D) -> Vector3D {
        let x: f32;
        let y: f32;
        let z: f32;

        x = if other.x == 0.0 { 0.0 } else { self.x / other.x };
        y = if other.y == 0.0 { 0.0 } else { self.y / other.y };
        z = if other.z == 0.0 { 0.0 } else { self.z / other.z };
        return Vector3D { x: x, y: y, z: z };
    }
}

impl ops::Mul<Vector3D> for Vector3D {
    type Output = Vector3D;

    fn mul(self, other: Vector3D) -> Vector3D {
        return Vector3D { x: (self.x * other.x), y: (self.z * other.y), z: (self.x * other.z) };
    }
}

impl ops::AddAssign<Vector3D> for Vector3D {

    fn add_assign(&mut self, other: Vector3D) {
        let result = *self + other;
        *self = result;
    }
}

impl ops::SubAssign<Vector3D> for Vector3D {

    fn sub_assign(&mut self, other: Vector3D) {
        let result = *self - other;
        *self = result;
    }
}

impl ops::MulAssign<Vector3D> for Vector3D {

    fn mul_assign(&mut self, other: Vector3D) {
        let result = *self * other;
        *self = result;
    }
}

impl ops::DivAssign<Vector3D> for Vector3D {

    fn div_assign(&mut self, other: Vector3D) {
        let result = *self / other;
        *self = result;
    }
}

impl ops::Mul<f32> for Vector3D {
    type Output = Vector3D;

    fn mul(self, value: f32) -> Vector3D {
        return Vector3D { x: self.x * value, y: self.y * value, z: self.z * value };
    }
}

impl ops::MulAssign<f32> for Vector3D {

    fn mul_assign(&mut self, value: f32) {
        let result = *self * value;
        *self = result;
    }
}

impl ops::Div<f32> for Vector3D {
    type Output = Vector3D;

    fn div(self, value: f32) -> Vector3D {
        return self * (1.0 / value);
    }
}

impl ops::DivAssign<f32> for Vector3D {

    fn div_assign(&mut self, value: f32) {
        let result = *self / value;
        *self = result;
    }
}
